File: basics/basics/topic_publisher.py
import rclpy
from rclpy.node import Node

from std_msgs.msg import Int32


class TopicPublisher(Node):

    def __init__(self):
        super().__init__('topic_publisher')
        self.publisher_ = self.create_publisher(Int32, 'counter', 10)
        timer_period = 0.5  # seconds
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.count = 0

    def timer_callback(self):
        msg = Int32()
        msg.data = self.count
        self.publisher_.publish(msg)
        self.count += 1


def main(args=None):
    rclpy.init(args=args)

    topic_publisher = TopicPublisher()

    rclpy.spin(topic_publisher)

    topic_publisher.destroy_node()
    rclpy.shutdown()


if __name__ == '__main__':
    main()

File: cpp_basics/src/topic_publisher.cpp
#include "rclcpp/rclcpp.hpp"
#include "wego_msgs/msg/counter.hpp"

#include <chrono>
#include <memory>
#include <functional>

using namespace std::chrono_literals;

class TopicPublisher : public rclcpp::Node
{
public:
    TopicPublisher()
    : Node("topic_publisher"), count_(0)
    {
        publisher_ = this->create_publisher<wego_msgs::msg::Counter>("counter", 10);
        
        auto timer_callback = [this](){ 
            auto message = wego_msgs::msg::Counter();
            message.count = this->count_;
            this->publisher_->publish(message);
            this->count_++;
        }; // 람다 식 사용 -> C++ 11 부터 auto는 반환값으로 자동 형변환

        timer_ = this->create_wall_timer(
                500ms, 
                timer_callback
        );
    }

private:
    rclcpp::TimerBase::SharedPtr timer_;
    rclcpp::Publisher<wego_msgs::msg::Counter>::SharedPtr publisher_;
    size_t count_;
};

int main(int argc, char ** argv)
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<TopicPublisher>());
    rclcpp::shutdown();
    return 0;
}
